### PR TITLE
Basic tap machines

### PR DESCRIPTION
**Pull request by d-rams**
This PR implements the ability to train binary RBMs by using direct gradient ascent using the TAP approximation for the Gibbs free energy term.

Check out the new example, `example_mnist_tap_machine.py` for a demo.

This is the "basic" PR because there are still a number of improvements to be made.

1. Use the 3rd order terms in the TAP expansion to make the free energy approximation more accurate
2. Investigate other means of minimizing the Legendre-transform of the free energy term other than simple gradient descent --which is what this implementation uses.
3. Investigate and implement dropout tap machines.

### CODE REVIEW

**Comment by d-rams** [comment]
Btw, the reason I removed backends.md was because it is backend-dependent. We should generate the docs perhaps for each and include both or neither... ?

**Comment by drckf** [comment]
There should be no difference between the two backends aside from the types in the signatures, so the docs are effectively the same. We could simply define some dummy types instead of using the built-ins.

**Comment by d-rams** [comment]
Yeah, I mean every time someone runs gendocs.py with a different backend, it will add a bunch of superfluous changes to the commit history.

**Comment by d-rams** [comment]
I will just manually keep the python-based signatures even though I'm using torch locally.
